### ISSUE TITLE
UltimateSDUpscale Error

### ISSUE THREAD

**Issue by bansengkwon** [ssitu/ComfyUI_UltimateSDUpscale#143]
I encountered this error while using ComfyUI.
I am using a computer with the following information:
System: macOS 15.3.2
Chip: Apple M2 Max
RAM: 32GB

<img width="1224" alt="Image" src="https://github.com/user-attachments/assets/ec27ff0a-9ae1-4e02-b9dd-d78277fd2c9b" />

<img width="266" alt="Image" src="https://github.com/user-attachments/assets/4e060110-de34-4caf-adbf-ac13e9e3a528" />

**Comment by phillipalsleben** [comment]
got the same issue

**Comment by ssitu** [comment]
Does it work on the cpu?

Likely related: https://github.com/ssitu/ComfyUI_UltimateSDUpscale/issues/128

**Comment by ssitu** [comment]
Looks like this is a PyTorch issue relating to Mac, and others have already discussed this in the ComfyUI repository.

You can either edit the code as shown here:
https://github.com/comfyanonymous/ComfyUI/issues/5075#issuecomment-2467393990

Or try changing your PyTorch version used by ComfyUI. [It seems like people had success with PyTorch 2.4.1](https://github.com/comfyanonymous/ComfyUI/issues/5075#issuecomment-2525080723). It might also be fixed in the PyTorch nightly 2.6 version if you'd like to try that instead.

**Comment by bansengkwon** [comment]
> Looks like this is a PyTorch issue relating to Mac, and others have already discussed this in the ComfyUI repository.看起来这是一个与 Mac 相关的 PyTorch 问题，其他人已经在 ComfyUI 仓库中讨论过这个问题了。
> 
> You can either edit the code as shown here:您可以在此处所示编辑代码： [comfyanonymous/ComfyUI#5075 (comment)](https://github.com/comfyanonymous/ComfyUI/issues/5075#issuecomment-2467393990)
> 
> Or try changing your PyTorch version used by ComfyUI. [It seems like people had success with PyTorch 2.4.1](https://github.com/comfyanonymous/ComfyUI/issues/5075#issuecomment-2525080723). It might also be fixed in the PyTorch nightly 2.6 version if you'd like to try that instead.尝试更改 ComfyUI 使用的 PyTorch 版本。似乎人们使用 PyTorch 2.4.1 取得了成功。如果您想尝试，PyTorch 的夜间 2.6 版本也可能已修复此问题。

I tried downgrading PyTorch to version 2.4.1, and it worked properly.

**Comment by bansengkwon** [comment]
> Looks like this is a PyTorch issue relating to Mac, and others have already discussed this in the ComfyUI repository.
> 
> You can either edit the code as shown here: [comfyanonymous/ComfyUI#5075 (comment)](https://github.com/comfyanonymous/ComfyUI/issues/5075#issuecomment-2467393990)
> 
> Or try changing your PyTorch version used by ComfyUI. [It seems like people had success with PyTorch 2.4.1](https://github.com/comfyanonymous/ComfyUI/issues/5075#issuecomment-2525080723). It might also be fixed in the PyTorch nightly 2.6 version if you'd like to try that instead.

I tried downgrading PyTorch to version 2.4.1, and it worked properly.

**Comment by VenimK** [comment]
I added the following lines to the utils.py file to fix the tensor view errors:

At the beginning of the tiled_scale_multidim function (line 871-872):
# Ensure input tensor is contiguous for Mac M1/M2 compatibility
samples = samples.contiguous()

Inside the loop where the function is applied to the tensor (around line 954):
# Ensure s_in is contiguous before applying the function for Mac M1/M2 compatibility
s_in = s_in.contiguous()

# Apply function and ensure result is contiguous
ps = function(s_in).contiguous().to(output_device)
This replaced the original line: ps = function(s_in).to(output_device)

When creating the mask (around line 957):
# Create mask and ensure it's contiguous for Mac M1/M2 compatibility
mask = torch.ones_like(ps).contiguous()
This replaced the original line: mask = torch.ones_like(ps)

In the final output calculation (around line 978):
# Ensure output tensors are contiguous for Mac M1/M2 compatibility
output[b:b+1] = (out/out_div).contiguous()

**Comment by ssitu** [comment]
I just pushed a potential fix, hopefully it fixes this problem. If anyone with a Mac can confirm whether it is still an issue or not, that would be great.

**Comment by GganallHis** [comment]
Unfortunately it didn't work for me the last fix :(
In my case I use the portable version in Windows (which works perfectly), and the desktop version of Confyui in MacOS (where the issue appears). Already tried to fix it downgrading pytorch, and also replacing the code in the RRDB.py file. 

FYI, tried the custom node of the Ultimate SD Upscale (GGUF) from quadmoon and works fine. Hope it helps!

**Comment by ssitu** [comment]
> > Could you please provide the full log of the error? It might give me some clues.
> 
> Sure, here you have it, thanks!
> 
> Also, apparently USDU (No Upscale) works fine. By plugging in the image upscaled with a model (then resized to the factor I wanted to upscale the original one), this node seems to tile and resample the image (externaly upscaled) fine. In case this helps.
> 
> [comfyui_8000.log](https://github.com/user-attachments/files/20443275/comfyui_8000.log)

Thank you for the log. I made another attempt to fix this issue. If you'd like to help me verify if it is fixed or not, run this command inside this custom node's directory: `git switch mps-noncontiguous-fix`. 
You can undo the changes by doing the following: `git switch main`.

**Comment by GganallHis** [comment]
Sorry, could you specify how to do it? It says "not a git repository" :/

**Comment by GganallHis** [comment]
> > Sorry, could you specify how to do it? It says "not a git repository" :/
> 
> Oh sorry, I assumed the custom node was downloaded a certain way. Try running the following in the custom nodes folder: `git clone -b mps-noncontiguous-fix https://github.com/ssitu/ComfyUI_UltimateSDUpscale.git`
> 
> This should download this custom node but with the new changes. Then I think you have to disable the original version to avoid conflicts.

Thanks! Sadly still not working, now says: "'ImageUpscaleWithModel' object has no attribute 'upscale_model'" (log attached).
But perhaps has nothing to do with your node, because I'm testing different upscale models. I uninstalled the fix and got back to your latest version. Works with 4xFFHQDAT.pth and 4xNomos8kHAT-L_otf.pth, doen't work with 4x_NMKD-Siax_200k.pth and 4x_NMKD-Superscale-SP_178000_G.pth, with whose I get the old error message. Same with the core node upscale using a model. 

[comfyui_8000_USDU2.log](https://github.com/user-attachments/files/20459603/comfyui_8000_USDU2.log)

**Comment by ssitu** [comment]
> Thanks! Sadly still not working, now says: "'ImageUpscaleWithModel' object has no attribute 'upscale_model'" (log attached). But perhaps has nothing to do with your node, because I'm testing different upscale models. I uninstalled the fix and got back to your latest version. Works with 4xFFHQDAT.pth and 4xNomos8kHAT-L_otf.pth, does'nt work with 4x_NMKD-Siax_200k.pth and 4x_NMKD-Superscale-SP_178000_G.pth, with whose I get the old error message. Same with the core node upscale using a model.
> 
> [comfyui_8000_USDU2.log](https://github.com/user-attachments/files/20459603/comfyui_8000_USDU2.log)

Sorry about that, it was a typo that I missed. I just pushed a fix for it if you want to try again, although it will not work if the core upscale model node doesn't work, because I reuse the core node inside this node for the upscaling.

**Comment by GganallHis** [comment]
> > Thanks! Sadly still not working, now says: "'ImageUpscaleWithModel' object has no attribute 'upscale_model'" (log attached). But perhaps has nothing to do with your node, because I'm testing different upscale models. I uninstalled the fix and got back to your latest version. Works with 4xFFHQDAT.pth and 4xNomos8kHAT-L_otf.pth, does'nt work with 4x_NMKD-Siax_200k.pth and 4x_NMKD-Superscale-SP_178000_G.pth, with whose I get the old error message. Same with the core node upscale using a model.
> > [comfyui_8000_USDU2.log](https://github.com/user-attachments/files/20459603/comfyui_8000_USDU2.log)
> 
> Sorry about that, it was a typo that I missed. I just pushed a fix for it if you want to try again, although it will not work if the core upscale model node doesn't work, because I reuse the core node inside this node for the upscaling. Did the GGUF version also work with 4x_NMKD-Siax_200k.pth and 4x_NMKD-Superscale-SP_178000_G.pth?

Here comes the weird thing.. All upscale models work in just one of the workflows I have, but these two are not in another one I tested and neither a new one I created just for this fix. I'll keep testing and will let you know if I find the clue. Thanks anyway for the effort and thanks a lot for creating this fantastic node :)

**Comment by Marx2** [comment]
https://github.com/comfyanonymous/ComfyUI/issues/5075#issuecomment-2451867268

Unfortunately this fix is outside of this node, but I can confirm it fix the problem on M1.
